### PR TITLE
[full-ci] More work on spaces support for sharing-ng

### PR DESCRIPTION
**Pull request by rhafer**
To add a space member:

```
curl --request POST \
  --url 'https://localhost:9200/graph/v1beta1/drives/0ec6cef0-9cd6-46a5-83ae-d4085a5f8a07$ae158e91-cb8b-45e0-ac9a-44ff5f6887d3/items/0ec6cef0-9cd6-46a5-83ae-d4085a5f8a07$ae158e91-cb8b-45e0-ac9a-44ff5f6887d3!ae158e91-cb8b-45e0-ac9a-44ff5f6887d3/invite' \
  --data '{
  "recipients": [
    {
      "@libre.graph.recipient.type": "user",
      "objectId": "4c510ada-c86b-4815-8820-42cdf82c3d51"
    }
  ],
  "roles": [ <roleid> ]
}'
```

gives the response:
```
{
  "value": [
    {
      "grantedToV2": {
        "user": {
          "displayName": "Albert Einstein",
          "id": "4c510ada-c86b-4815-8820-42cdf82c3d51"
        }
      },
      "id": "u:4c510ada-c86b-4815-8820-42cdf82c3d51",
      "roles": [
        "a8d5fe5e-96e3-418d-825b-534dbdf22b99"
      ]
    }
  ]
}
```

To remove a permission:
```
curl --request DELETE \
  --url 'https://localhost:9200/graph/v1beta1/drives/0ec6cef0-9cd6-46a5-83ae-d4085a5f8a07$ae158e91-cb8b-45e0-ac9a-44ff5f6887d3/items/0ec6cef0-9cd6-46a5-83ae-d4085a5f8a07$ae158e91-cb8b-45e0-ac9a-44ff5f6887d3!ae158e91-cb8b-45e0-ac9a-44ff5f6887d3/permissions/u:4c510ada-c86b-4815-8820-42cdf82c3d51'
```

The permission-id (last part of the URL) can be taken from the `GET https://localhost:9200/graph/v1beta1/drives/{{driveid}}/items/{{itemid}}/permissions` response.

This needs: https://github.com/cs3org/reva/pull/4585

### CODE REVIEW

**Comment by rhafer** [comment]
> This needs: [cs3org/reva#4585](https://github.com/cs3org/reva/pull/4585)

I can't currently bump reva because, ocis will not build with current edge until https://github.com/owncloud/ocis/pull/8701 is done.

However this PR should work with out the reva changes. It will just lack the check for having a least one space manager remaining when changing/delete permissions via graph.

**Comment by kobergj** [comment]
> I can't currently bump reva because, ocis will not build with current edge until https://github.com/owncloud/ocis/pull/8701 is done.

@rhafer why do you think this? `edge` builds with current `master`. Or you just linked the wrong PR? https://github.com/owncloud/ocis/pull/8701 is not fixing anything.

File: services/graph/pkg/service/v0/driveitems.go [hunk 1]
```diff
@@ -691,15 +684,29 @@ func (g Graph) DeletePermission(w http.ResponseWriter, r *http.Request) {
...
+		err = nil
+	}
+
+	// If this is neither a public share nor a space permission, check if this is a
+	// user share
+	if permissionType == Unknown {
+		sharedResourceID, err = g.getUserPermissionResourceID(ctx, permissionID)
+		if err == nil {
+			permissionType = User
+		}
```
Comment (kobergj) [inline services/graph/pkg/service/v0/driveitems.go:709]:
```suggestion
	sharedResourceID, err := g.getLinkPermissionResourceID(ctx, permissionID)
	switch {
	// Check if the ID is referring to a public share
	case err == nil:
		permissionType = Public

	// If the item id is referring to a space root an this is not a public share
	// we have to deal with space permissions
	case IsSpaceRoot(&itemID):
		permissionType = Space
		sharedResourceID = &itemID

	// If this is neither a public share nor a space permission, check if this is a
	// user share
	default:
		sharedResourceID, err = g.getUserPermissionResourceID(ctx, permissionID)
		if err == nil {
			permissionType = User
		}
	}
```

Wouldn't this be simpler with a switch case? Or am I misunderstanding the logic? :smile:

File: services/graph/pkg/service/v0/driveitems.go [hunk 2]
```diff
@@ -731,34 +741,71 @@ func (g Graph) DeletePermission(w http.ResponseWriter, r *http.Request) {
...
+		if err != nil {
+			g.logger.Debug().Err(err).Msg("selecting gatewaySelector failed")
+			return nil, nil, err
+		}
+		// get space id
+		statResponse, err := gatewayClient.Stat(ctx, &storageprovider.StatRequest{Ref: &storageprovider.Reference{ResourceId: itemID}})
+		if errCode := errorcode.FromStat(statResponse, err); errCode != nil {
+			g.logger.Warn().Err(errCode).Interface("stat.res", statResponse).Msg("stat failed")
+			return nil, nil, errCode
+		}
```
Comment (kobergj) [inline services/graph/pkg/service/v0/driveitems.go:767]: You can use `utils.GetResourceByID` if you don't want to hassle with stat calls. (not critical ofc)

File: services/graph/pkg/service/v0/driveitems.go [hunk 3]
```diff
@@ -731,34 +741,71 @@ func (g Graph) DeletePermission(w http.ResponseWriter, r *http.Request) {
...
 
 	return nil, nil, err
 
 }
 
+func (g Graph) getSpaceRootPermissions(ctx context.Context, spaceID *storageprovider.StorageSpaceId) ([]libregraph.Permission, error) {
+	filters := []*storageprovider.ListStorageSpacesRequest_Filter{
+		listStorageSpacesIDFilter(spaceID.GetOpaqueId()),
+	}
+	res, err := g.ListStorageSpacesWithFilters(ctx, filters, true)
```
Comment (kobergj) [inline services/graph/pkg/service/v0/driveitems.go:801]: There is also `utils.GetSpace`. Just saying....

**Review by kobergj — CHANGES_REQUESTED** [review]
Just some details :+1:

**Comment by rhafer** [comment]
> @rhafer why do you think this? `edge` builds with current `master`. Or you just linked the wrong PR? #8701 is not fixing anything.

IIRC it failed building against edge. Complained about some issues regarding initiatorIDs. I don't remember the exact details. And it works now so :shrug:

File: services/graph/pkg/service/v0/driveitems.go (hunk 2 above)
Comment (rhafer) [reply services/graph/pkg/service/v0/driveitems.go:767]: Hm, nice

File: services/graph/pkg/service/v0/driveitems.go (hunk 3 above)
Comment (rhafer) [reply services/graph/pkg/service/v0/driveitems.go:801]: Hm, nice²

File: services/graph/pkg/service/v0/driveitems.go (hunk 1 above)
Comment (rhafer) [reply services/graph/pkg/service/v0/driveitems.go:709]: I guess you're right. Seems I stared at this code for too long.

**Review by kobergj — APPROVED** [review]

**Pull request merged**